Packages/com.unity.inputsystem/InputSystem/InputManager.cs

```diff
@@ -1207,6 +1207,16 @@ public InputDevice AddDevice(InputDeviceDescription description)
             return device;
         }
 
+        public InputDevice AddDevice(InputDeviceDescription description, InternedString layout, string deviceName = null,
+            int deviceId = InputDevice.InvalidDeviceId, InputDevice.DeviceFlags deviceFlags = 0)
+        {
+            Profiler.BeginSample("InputSystem.AddDevice");
```

Comment:
Since I noticed that AddDevice is called within try-catch below, does the Profiler.BeginSample() / Profiler.EndSample() pair need to a scoped RAII operation, e.g. using (new ScopedProfilerSample()) { ..... } in order to guarantee that EndSample() is called also if AddDevice throws? If AddDevice never throws, is the try-catch needed?

Reply:
It can throw further down the call tree. I'm not familiar with ScopedProfilerSample. We probably should be using ProfilerMarker here and Auto for scoping, but that's the newer profiling API and I'm not sure if there are any complications from switching to it.

Reply:
There is no ScopedProfilerSample as far as I know it was more of "psuedo description" but what I meant was roughly (borrowing form C++)

`
class ScopedProfilerSample : IDisposable
{
    ScopedProfilerSample(string id) { this.id = id; profiler.BeginSample("InputSystem.AddDevice"); }
    void Dispose() { profiler.EndSample(id); }
    private string id;
}
`

So you could use it as a scoped guard in a "using" statement (as all other RAII-like stuff with using statements)

Reply:
Andrew I think it would make sense to ensure that Profiler.EndSample() is called also if AddDevice throws if AddDevice may throw. However, the description I assume should not be updated in that case so probably just EndSample in catch -> rethrow while the Add + description setter remains in "try" part?